Introduce ProfileCollectorManager

cbuescher opened the pull request:
Spin off of #96689 that only contains the new collector manager that wraps all 
Collectors generated by it in an InternalProfileCollector.

cbuescher committed:
Introduce MultiProfileCollectorManager

Iter, adding tests

iter

fix docs test

wip iter

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

cbuescher committed:
Merge branch 'main' into intro-MultiProfileCollectorManager

cbuescher committed:
Add test for multi-collector profiling

javanna commented on server/src/main/java/org/elasticsearch/search/profile/query/ProfileCollectorManager.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <C extends Collector> ProfileCollectorManager create(CollectorManager<C, ?> collectorManager) {
```
you could do `? extends Collector` in the constructor?

javanna commented on server/src/main/java/org/elasticsearch/search/profile/query/ProfileCollectorManager.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Override
+    public InternalProfileCollector newCollector() throws IOException {
+        return new InternalProfileCollector(collectorManager.newCollector(), CollectorResult.REASON_SEARCH_TOP_HITS);
```
can you make the reason configurable instead of hardcoded?

javanna replied:
Also, in #96622 I added support for setting the child collectors as well. Could you double check if that makes sense and integrate it?

cbuescher replied:
I added the "reason" to the constructor. As for the child collectors, I can see where this will probably be needed later in your PR but I would prefer to add it there, so the motivation is clear.

javanna commented on server/src/test/java/org/elasticsearch/search/profile/query/ProfileCollectorManagerTests.java:
```diff
@@ -0,0 +1,155 @@
+        }
+    }
+
+    public void testNewCollector() throws IOException {
```
I am thinking that testing newCollector and reduce separately is a bit overkill, if we test a search with an index searcher that calls them both like we do below. What do you think?

cbuescher replied:
I joined the two first tests into one and removed the getResultTree() test in favour of the one with the index searcher. I think its still good to have one "stupid" test just for newCollector() and reduce(). It communicates its intent more clearly. I also added javadocs to the tests because I often find it hard to remember the intent of these tests later.

javanna reviewed (changes requested):
Thanks for working on this, left some comments.

cbuescher committed:
adressing review comments

cbuescher committed:
Merge branch 'main' into intro-MultiProfileCollectorManager

cbuescher commented:
@javanna I pushed some changes after your last review.
In particular I also changed the parent CollectorResult a bit, now the "reason" is "parent_collector_manager" with a timer of 0 to indicate that the collector manager itself doesn't do much. I removed the "max" of the child collectors, I'd also not be in favor of "sum" though. Maybe at some point later the end-to-end timing of running the threads can be captured here though.
```
{
  "name" : "ProfileCollectorManager",
  "reason" : "parent_collector_manager",
  "time_in_nanos" : 0,
  "children" : [
    {
      "name" : "SimpleTopScoreDocCollector",
      "reason" : "profiler_reason",
      "time_in_nanos" : 186292
    }
  ]

```

javanna reviewed (approved):
left a couple more thoughts, LGTM though, up to you where you want to add support for the children collectors.

cbuescher commented on server/src/main/java/org/elasticsearch/search/profile/query/ProfileCollectorManager.java:
```diff
@@ -0,0 +1,60 @@
+
+    @SuppressWarnings("unchecked")
+    public ProfileCollectorManager(CollectorManager<? extends Collector, ?> collectorManager, String reason) {
+        this.collectorManager = (CollectorManager<Collector, ?>) collectorManager;
```
That causes issues when we call reduce() in L45 with the unwrapped collectors. I don't fully understand why but I was fiddling with it for some time. I'll leave it like this for now for the sake of making progress here.

cbuescher commented on server/src/test/java/org/elasticsearch/search/profile/query/ProfileCollectorManagerTests.java:
```diff
@@ -0,0 +1,118 @@
+
+            CollectorResult parent = profileCollectorManager.getCollectorTree();
+            assertEquals("ProfileCollectorManager", parent.getName());
+            assertEquals("parent_collector_manager", parent.getReason());
```
Naming is hard. I like your suggestion, but that assumes search is concurrent. I was under the impression that we'd also see the single-threaded case quite often still? I'm going ahead and call it "segment_search" instead, its not hinting at the implementation details (ie collector manager) too much but also doesn't assume anything about the mode of execution.

cbuescher committed:
small renaming of parent collector result reason
